src/Screens/Flows/App/NotificationScreen/NotificationScreen.tsx

```diff
@@ -110,14 +111,62 @@ export const NotificationScreen = () => {
                     return
                 }
 
-                addTag(tag, "true")
+                // Prevent rapid toggling
+                if (isUpdatingTags[tag]) return
+
+                // Set loading state
+                setIsUpdatingTags(prev => ({ ...prev, [tag]: true }))
+
+                // Optimistic update
+                setTags(prev => ({ ...prev, [tag]: "true" }))
+
+                try {
+                    addTag(tag, "true")
+                } catch (apiError) {
+                    // Revert optimistic update on error
+                    setTags(prev => {
+                        const newTags = { ...prev }
+                        delete newTags[tag]
+                        return newTags
+                    })
+                } finally {
+                    // Clear loading state
+                    setIsUpdatingTags(prev => {
+                        const newState = { ...prev }
+                        delete newState[tag]
+                        return newState
+                    })
+                }
             } else {
-                removeTag(tag)
+                // Prevent rapid toggling
+                if (isUpdatingTags[tag]) return
+
+                // Set loading state
+                setIsUpdatingTags(prev => ({ ...prev, [tag]: true }))
+
+                // Optimistic update
+                setTags(prev => {
+                    const newTags = { ...prev }
+                    delete newTags[tag]
+                    return newTags
+                })
+
+                try {
+                    removeTag(tag)
+                } catch (apiError) {
+                    // Revert optimistic update on error
+                    setTags(prev => ({ ...prev, [tag]: "true" }))
+                } finally {
+                    // Clear loading state
+                    setIsUpdatingTags(prev => {
+                        const newState = { ...prev }
+                        delete newState[tag]
+                        return newState
+                    })
+                }
```

Comment:
I believe all this logic could be simplified by making `updateTags` return the Promise
And wrap it in a useQuery